### DISCUSSION TITLE
OpenAPI docs fails (sometimes) to load when using nested discriminator

### DISCUSSION THREAD

**Discussion by tarekasishm** [fastapi/fastapi#10169] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Literal, Union
from typing_extensions import Annotated

from fastapi import FastAPI
from pydantic import BaseModel, Field
import uvicorn

class BlackCat(BaseModel):
    pet_type: Literal['cat']
    color: Literal['black']
    black_name: str


class WhiteCat(BaseModel):
    pet_type: Literal['cat']
    color: Literal['white']
    white_name: str


Cat = Annotated[Union[BlackCat, WhiteCat], Field(discriminator='color')]


class Dog(BaseModel):
    pet_type: Literal['dog']
    name: str

class Model(BaseModel):
    pet: Union[Cat, Dog] = Field(..., discriminator="pet_type")
    n: int

app = FastAPI()

@app.get("/")
def controller(model: Model) -> str:
    return "Ok!"

uvicorn.run(app, port=8000)
```

### Description

* Just try to open the OpenAPI Doc (`http://127.0.0.1:8000/docs`) and it will fail to load the model

### Operating System

Linux

### FastAPI Version

0.103.0

### Pydantic Version

2.2.1

### Python Version

Python 3.11.4

### Additional Context

I think it's related to the model validation that `FastAPI` performs (although it's a valid pydantic model) because I could get generated `OpenAPI` schema and it's a valid one. Additionally, if you define pet as a list instead o flat union, everything works as expected:
```python
class Model(BaseModel):
    pet: list[Union[Cat, Dog]] = Field(..., discriminator="pet_type")
    n: int
```

**Comment by JavierSanchezCastro** [comment]
Actually, the Discriminator class doesn't allow having a nested Schema in `mapping` property, see here: https://github.com/tiangolo/fastapi/blob/master/fastapi/openapi/models.py#L136

It also matches with openapi spec: https://spec.openapis.org/oas/v3.1.0#fixed-fields-20

**Reply by JavierSanchezCastro** [reply]
But for this example the discriminator by color is weird, you can use an enum for that

```python3
class CatColor(Enum):
    BLACK = "black"
    WHITE = "white"

class Cat(BaseModel):
    pet_type: Literal['cat']
    color: CatColor
    name: str

class Dog(BaseModel):
    pet_type: Literal['dog']
    name: str

Pet = Annotated[Cat | Dog, Field(discriminator='pet_type')]

class Model(BaseModel):
    pet: Pet
    n: int
```

**Reply by tarekasishm** [reply]
Take it just as a minimal example. I found this behavior in a larger base code but this snippet lets you reproduce it and you can focus on the relevant parts.  

Regarding the OpenAPI spec, I'm not sure if what you say is accurate, This document just defines what a discriminator is and how to use it but I can't see anything related to nested discriminator. Additionally, *Redoc* supports nested discriminators and the OpenAPI spec generated for this schema is compliant with the `3.1.0` version. I've used `openapi-spec-validator` to check it and I'll attach the spec in this comment.
[openapi.zip](https://github.com/tiangolo/fastapi/files/12475188/openapi.zip)

**Reply by sarisia** [reply]
I'm facing the same.

Also,

1. The example workaround @tarekasishm gave in `Additional Context` doesn't work. It raises:

```
Traceback (most recent call last):
  File "/workspaces/xxx-api/test.py", line 34, in <module>
    class Model(BaseModel):
  File "pydantic/main.py", line 197, in pydantic.main.ModelMetaclass.__new__
  File "pydantic/fields.py", line 506, in pydantic.fields.ModelField.infer
  File "pydantic/fields.py", line 436, in pydantic.fields.ModelField.__init__
  File "pydantic/fields.py", line 552, in pydantic.fields.ModelField.prepare
  File "pydantic/fields.py", line 633, in pydantic.fields.ModelField._type_analysis
TypeError: `discriminator` can only be used with `Union` type with more than one variant
```

2. It seems that both of Redoc and swagger-ui doesn't have support for nested discriminator. Both cannot render the spec attached as `openapi.zip` correctly and raises... (tested with [Redoc Demo](https://redocly.github.io/redoc/) and [Swagger Editor](https://editor-next.swagger.io/))
